### PR TITLE
[vcpkg] Add vcpkg_from_sourceforge (1/2)

### PR DESCRIPTION
**Pull request by JackBoosY**
There are 43 ports in vcpkg downloaded using sourceforge.
Adding a unified download method is convenient for modifying the download method of the sourceforge storage port in the future.

Related: #11502.

- [x] Add function vcpkg_from_sourceforge
- [x] Modify all ports involved in download
- [x] Add documentation

### CODE REVIEW

**Review by BillyONeal — APPROVED** [review]

**Comment by JackBoosY** [comment]
@Neumann-A @Voskrese @cenit Could you please review this PR?

Thanks.

File: docs/maintainers/vcpkg_from_sourceforge.md [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+Download and extract a project from sourceforge.
+
+## Usage:
+```cmake
+vcpkg_from_sourceforge(
+    OUT_SOURCE_PATH <SOURCE_PATH>
+    REPO <cunit/CUnit>
+    [REF] <2.1-3>
+    SHA512 <547b417109332...>
+    FILENAME <CUnit-2.1-3.tar.bz2>
```
Comment (cenit) [inline docs/maintainers/vcpkg_from_sourceforge.md:12]: FILENAME is not documented. For example, for newbies it would be so clear if it's the name of the archive on the server or on the local computer, or both...

**Comment by cenit** [comment]
apart from one single note, it's a very nice addition IMHO.
Thanks @JackBoosY

File: docs/maintainers/vcpkg_from_sourceforge.md (hunk 1 above)
Comment (JackBoosY) [reply docs/maintainers/vcpkg_from_sourceforge.md:12]: Because of this, there are specific examples below.
See line 31.

Comment (cenit) [reply docs/maintainers/vcpkg_from_sourceforge.md:12]: Yes I agree, but still there is no `### FILENAME` anywhere in the docs

**Comment by JackBoosY** [comment]
In the next PR, I will add code to retry different mirrors in this function.

File: ports/irrlicht/portfile.cmake [hunk 2]
```diff
@@ -1,17 +1,12 @@
 vcpkg_fail_port_install(ON_ARCH "arm" ON_TARGET "osx" "uwp")
 
-vcpkg_download_distfile(ARCHIVE
-    URLS "https://downloads.sourceforge.net/project/irrlicht/Irrlicht%20SDK/1.8/1.8.4/irrlicht-1.8.4.zip"
+vcpkg_from_sourceforge(
+    OUT_SOURCE_PATH SOURCE_PATH
+    REPO irrlicht/Irrlicht%20SDK
```
Comment (PhoebeHui) [inline ports/irrlicht/portfile.cmake:5]: For me, the %20 in repo may confuse to users, it will be good if give an example to users.

File: ports/libaiff/portfile.cmake [hunk 3]
```diff
@@ -1,17 +1,14 @@
 vcpkg_check_linkage(ONLY_STATIC_LIBRARY)
 
-vcpkg_download_distfile(ARCHIVE
-    URLS "https://sourceforge.net/projects/aifftools/files/libaiff/LibAiff%205.0/libaiff-5.0-release.tar.gz"
+vcpkg_from_sourceforge(
+    OUT_SOURCE_PATH SOURCE_PATH
+    REPO aifftools/libaiff
+    REF LibAiff%205.0
```
Comment (PhoebeHui) [inline ports/libaiff/portfile.cmake:6]: Same as above.

File: scripts/cmake/vcpkg_from_sourceforge.cmake [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+    
+    if (DEFINED _vdus_REF)
+        set(URL "${SOURCEFORGE_HOST}/${ORG_NAME}${REPO_NAME}/${_vdus_REF}/${_vdus_FILENAME}")
+    else()
+        set(URL "${SOURCEFORGE_HOST}/${ORG_NAME}${REPO_NAME}/${_vdus_FILENAME}")
+    endif()
+        
+    set(NO_REMOVE_ONE_LEVEL )
+    if (_vdus_NO_REMOVE_ONE_LEVEL)
+        set(NO_REMOVE_ONE_LEVEL NO_REMOVE_ONE_LEVEL)
```
Comment (dan-shaw) [inline scripts/cmake/vcpkg_from_sourceforge.cmake:112]: Is this supposed to be set to `_vdus_NO_REMOVE_ONE_LEVEL` instead?

File: scripts/cmake/vcpkg_from_sourceforge.cmake [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+    if (NOT FOUND_ORG EQUAL -1)
+        string(REGEX REPLACE ".*/" "" REPO_NAME ${_vdus_REPO})
+        string(REGEX REPLACE "/.*" "" ORG_NAME ${_vdus_REPO})
+        set(ORG_NAME ${ORG_NAME}/)
+    else()
+        set(REPO_NAME ${_vdus_REPO})
+        set(ORG_NAME )
+    endif()
+
+    if(VCPKG_USE_HEAD_VERSION AND NOT DEFINED _vdus_HEAD_REF)
```
Comment (dan-shaw) [inline scripts/cmake/vcpkg_from_sourceforge.cmake:99]: Is `_vdus_HEAD_REF` set somewhere? Otherwise, I think `cmake_parse_arguments` needs to parse this.

**Review by dan-shaw — APPROVED** [review]

File: scripts/cmake/vcpkg_from_sourceforge.cmake (hunk 5 above)
Comment (JackBoosY) [reply scripts/cmake/vcpkg_from_sourceforge.cmake:99]: Ops, this should be deleted.

File: ports/irrlicht/portfile.cmake (hunk 2 above)
Comment (JackBoosY) [reply ports/irrlicht/portfile.cmake:5]: Example was already written in the [document](https://github.com/microsoft/vcpkg/pull/11899/files?file-filters%5B%5D=.cmake&file-filters%5B%5D=.md&file-filters%5B%5D=.txt&file-filters%5B%5D=No+extension#diff-023f4e1b7e79623fa98321695b8b227dR34).

File: ports/libaiff/portfile.cmake (hunk 3 above)
Comment (JackBoosY) [reply ports/libaiff/portfile.cmake:6]: Same as above.

File: ports/irrlicht/portfile.cmake (hunk 2 above)
Comment (JackBoosY) [reply ports/irrlicht/portfile.cmake:5]: The `%20` is the string after being transcoded using url transcoding.

**Comment by JackBoosY** [comment]
`libpq:arm64-windows` regression will be fixed in #11965.

**Comment by JackBoosY** [comment]
Done.

File: scripts/cmake/vcpkg_from_sourceforge.cmake (hunk 5 above)
Comment (JackBoosY) [reply scripts/cmake/vcpkg_from_sourceforge.cmake]: Non-versioned repo does not support the head option.

File: scripts/cmake/vcpkg_from_sourceforge.cmake (hunk 4 above)
Comment (cenit) [reply scripts/cmake/vcpkg_from_sourceforge.cmake]: in any case `set(NO_REMOVE_ONE_LEVEL NO_REMOVE_ONE_LEVEL)` looks not right...

Comment (JackBoosY) [reply scripts/cmake/vcpkg_from_sourceforge.cmake]: @cenit Therefore, I added double quotes to it.

File: docs/maintainers/vcpkg_from_sourceforge.md [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+### SHA512
+The SHA512 hash that should match the archive.
+
+### PATCHES
+A list of patches to be applied to the extracted sources.
+
+Relative paths are based on the port directory.
+
+This field should contain the scheme, host, and port of the desired URL without a trailing slash.
```
Comment (strega-nil) [inline docs/maintainers/vcpkg_from_sourceforge.md:50]: This doesn't seem related to `PATCHES`; aren't they normally local?

File: scripts/cmake/vcpkg_from_sourceforge.cmake [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+
+    if (_vdus_DISABLE_SSL)
+        set(URL_PROTOCOL http:)
+    else()
+        set(URL_PROTOCOL https:)
+    endif()
+    set(SOURCEFORGE_HOST ${URL_PROTOCOL}//downloads.sourceforge.net/project)
+
+    string(FIND ${_vdus_REPO} "/" FOUND_ORG)
+    if (NOT FOUND_ORG EQUAL -1)
```
Comment (strega-nil) [inline scripts/cmake/vcpkg_from_sourceforge.cmake:90]: Would it be possible for you to print an explicit error if there are multiple `/`s in the string?

File: scripts/cmake/vcpkg_from_sourceforge.cmake [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+        string(REGEX REPLACE ".*/" "" REPO_NAME ${_vdus_REPO})
```
Comment (strega-nil) [inline scripts/cmake/vcpkg_from_sourceforge.cmake:91]: Can you use `string(SUBSTRING` instead here? something like

```cmake
string(SUBSTRING "${_vdus_REPO}" 0 FOUND_ORG ORG_NAME)
string(SUBSTRING "${_vdus_REPO}" FOUND_ORG -1 REPO_NAME)
```

**Review by strega-nil — CHANGES_REQUESTED** [review]
Some minor comments :)

File: scripts/cmake/vcpkg_from_sourceforge.cmake (hunk 7 above)
Comment (JackBoosY) [reply scripts/cmake/vcpkg_from_sourceforge.cmake:90]: @strega-nil `/` in `REPO` will only have 0 or 1, this is decided by sourceforge.

Comment (JackBoosY) [reply scripts/cmake/vcpkg_from_sourceforge.cmake:90]: For example:
```
https://sourceforge.net/settings/mirror_choices?projectname=cunit&filename=CUnit/2.1-3/CUnit-2.1-3.tar.bz2&selected=nchc
https://sourceforge.net/settings/mirror_choices?projectname=argtable&filename=argtable/argtable-2.13/argtable2-13.tar.gz&selected=nchc
https://sourceforge.net/settings/mirror_choices?projectname=geographiclib&filename=distrib/archive/GeographicLib-1.47-patch1.zip&selected=ayera
https://sourceforge.net/settings/mirror_choices?projectname=giflib&filename=giflib-5.1.9.tar.bz2&selected=astuteinternet
https://sourceforge.net/settings/mirror_choices?projectname=gsoap2&filename=gsoap-2.8/gsoap_2.8.103.zip&selected=jaist
https://sourceforge.net/settings/mirror_choices?projectname=mad&filename=libmad/0.15.1b/libmad-0.15.1b.tar.gz&selected=nchc
https://sourceforge.net/settings/mirror_choices?projectname=libsquish&filename=libsquish-1.15.tgz&selected=nchc
https://sourceforge.net/settings/mirror_choices?projectname=libuuid&filename=libuuid-1.0.3.tar.gz&selected=nchc
https://sourceforge.net/settings/mirror_choices?projectname=polyclipping&filename=clipper_ver6.4.2.zip&selected=nchc
https://sourceforge.net/settings/mirror_choices?projectname=soxr&filename=soxr-0.1.3-Source.tar.xz&selected=nchc
```

File: scripts/cmake/vcpkg_from_sourceforge.cmake (hunk 8 above)
Comment (JackBoosY) [reply scripts/cmake/vcpkg_from_sourceforge.cmake:91]: I think that would be okay, this code from [vcpkg_from_github.cmake](https://github.com/microsoft/vcpkg/blob/ee17a685087a6886e5681e355d36cd784f0dd2c8/scripts/cmake/vcpkg_from_github.cmake#L103).
And if we use SUBSTRING, we need to deal with additional `/`.

File: scripts/cmake/vcpkg_from_sourceforge.cmake (hunk 7 above)
Comment (strega-nil) [reply scripts/cmake/vcpkg_from_sourceforge.cmake:88]: right, but I want it to print an error if you accidentally put two slashes (for example, someone might put `x/y/`

File: scripts/cmake/vcpkg_from_sourceforge.cmake (hunk 8 above)
Comment (strega-nil) [reply scripts/cmake/vcpkg_from_sourceforge.cmake:89]: Yeah, I think we should also probably fix `vcpkg_from_github` to use `string(SUBSTRING`

File: docs/maintainers/vcpkg_from_sourceforge.md [hunk 9]
```diff
@@ -0,0 +1,63 @@
+# vcpkg_from_sourceforge
+
+Download and extract a project from sourceforge.
+
+## Usage:
+```cmake
+vcpkg_from_sourceforge(
+    OUT_SOURCE_PATH <SOURCE_PATH>
```
Comment (strega-nil) [inline docs/maintainers/vcpkg_from_sourceforge.md:8]: If this is almost always going to be `SOURCE_PATH`, is there any reason not to make the parameter optional and default it to `SOURCE_PATH`?

File: scripts/cmake/vcpkg_from_sourceforge.cmake [hunk 10]
```diff
@@ -0,0 +1,125 @@
+## # vcpkg_from_sourceforge
+##
+## Download and extract a project from sourceforge.
+##
+## ## Usage:
+## ```cmake
+## vcpkg_from_sourceforge(
+##     OUT_SOURCE_PATH <SOURCE_PATH>
+##     REPO <cunit/CUnit>
+##     [REF] <2.1-3>
```
Comment (strega-nil) [inline scripts/cmake/vcpkg_from_sourceforge.cmake:10]:
```suggestion
##     [REF <2.1-3>]
```

**Review by strega-nil — COMMENTED** [review]
That's the only things I see; I would still like to see moving from regex to substring, and printing an error if there's more than one `/`

File: scripts/cmake/vcpkg_from_sourceforge.cmake [hunk 11]
```diff
@@ -86,8 +86,16 @@ function(vcpkg_from_sourceforge)
...
+        # Check for the second slash
+        math(EXPR NEXT_SLASH "${FOUND_ORG} + 1")
+        string(SUBSTRING ${_vdus_REPO} ${NEXT_SLASH} -1 LAST_STRING)
+        string(FIND ${LAST_STRING} "/" FOUND_NEXT_SLASH)
+        if (NOT FOUND_NEXT_SLASH EQUAL -1)
+            message(FATAL_ERROR "REPO should contain at most one slash.\nExample: mad/libmad")
+        endif()
+        
+        string(SUBSTRING "${_vdus_REPO}" 0 FOUND_ORG ORG_NAME)
+        string(SUBSTRING "${_vdus_REPO}" FOUND_ORG -1 REPO_NAME)
```
Comment (strega-nil) [inline scripts/cmake/vcpkg_from_sourceforge.cmake:98]: Might be better to do something like:

```cmake
string(SUBSTRING "${_vdus_REPO}" 0 FOUND_ORG ORG_NAME)
string(SUBSTRING "${_vdus_REPO}" FOUND_ORG -1 REPO_NAME)
if (REPO_NAME MATCHES "/")
    message(FATAL_ERROR "REPO should contain at most one slash (found ${_vdus_REPO}).")
endif()
```

then there's not quite as much code.

**Review by BillyONeal — APPROVED** [review]

**Review by strega-nil — APPROVED** [review]
Thanks JacK!

**Comment by strega-nil** [comment]
What are the blockers to merging this? Just CI running?

**Comment by vicroms** [comment]
Just CI

**Review by vicroms — APPROVED** [review]

**Comment by BillyONeal** [comment]
Yeah, for some reason the previous attempt broke inside Azure Pipelines (some of the machines dropped off the face of the earth)

**Comment by strega-nil** [comment]
Cool :)

File: ports/pcre2/portfile.cmake [hunk 12]
```diff
@@ -1,17 +1,15 @@
 set(PCRE2_VERSION 10.30)
-include(vcpkg_common_functions)
 
-vcpkg_download_distfile(ARCHIVE
-    URLS "https://ftp.pcre.org/pub/pcre/pcre2-${PCRE2_VERSION}.zip" "https://sourceforge.net/projects/pcre/files/pcre2/${PCRE2_VERSION}/pcre2-${PCRE2_VERSION}.zip/download"
```
Comment (c72578) [inline ports/pcre2/portfile.cmake:5]: In case of pcre2, the primary download server was:
https://ftp.pcre.org/pub/pcre/
This has been removed here by switching to `vcpkg_from_sourceforge()`.
The following appears now in artifacts of failed CI-builds [1]:
```console
-- Downloading https://downloads.sourceforge.net/project/pcre/pcre2/10.30/pcre2-10.30.zip... Failed. Status: 7;"Couldn't connect to server"
```
During the CI-build of commit a0bd59c6a278f3c173f3a8cf802c5148b6489ff6 it occurred two times (pcre2:x64-windows and pcre2:x64-uwp).
Is there a possibility to have a different download server like https://ftp.pcre.org/pub/pcre/ also when using `vcpkg_from_sourceforge()`? Otherwise the previous `vcpkg_download_distfile()` used for pcre2 would be more reliable.

[1] https://dev.azure.com/vcpkg/public/_build/results?buildId=39415&view=results

Comment (JackBoosY) [reply ports/pcre2/portfile.cmake:5]: I will take a look later.

Comment (JackBoosY) [reply ports/pcre2/portfile.cmake:5]: @c72578 Try to downloads source with all mirrors has been added in my other PR #12018. You can try that.

Comment (c72578) [reply ports/pcre2/portfile.cmake:5]: @JackBoosY Thanks, that's exactly what is needed.

Comment (BillyONeal) [reply ports/pcre2/portfile.cmake:5]: @c72578 Does https://github.com/microsoft/vcpkg/pull/12233 do what you want?

Comment (c72578) [reply ports/pcre2/portfile.cmake:5]: @BillyONeal, yes #12233 LGTM. It primarily uses the original download server and as an alternative possibility the improved download from sourceforge.
